python-client/giskard/models/model_explanation.py

```diff
@@ -6,46 +6,114 @@
-    feature_names = list(df.columns)
+    # Select SHAP values, which explain model's predictions with the highest probability.
+    filtered_shap_values = list()
+    for sample_idx, sample_prediction_idx in enumerate(predictions):
+        prediction_explanation = shap_values[sample_prediction_idx][sample_idx]
+        filtered_shap_values.append(prediction_explanation)
+
+    return filtered_shap_values
+
+
+def prepare_df(df: pd.DataFrame, model: BaseModel, dataset: Dataset) -> pd.DataFrame:
+    """Prepare dataframe for an inference step."""
+    df = model.prepare_dataframe(df, column_dtypes=dataset.column_dtypes, target=dataset.target)
+
+    if dataset.target in df.columns:
+        prepared_dataset = Dataset(df, column_types=dataset.column_types, target=dataset.target)
+    else:
+        prepared_dataset = Dataset(df, column_types=dataset.column_types)
+
+    # Make sure column order is the same as in the dataset.df.
+    columns_original_order = (
+        model.meta.feature_names
+        if model.meta.feature_names
+        else [c for c in dataset.df.columns if c in prepared_dataset.df.columns]
+    )
+
+    prepared_df = prepared_dataset.df[columns_original_order]
+    return prepared_df
+
+
+def background_example(df: pd.DataFrame, input_types: Dict[str, str]) -> pd.DataFrame:
+    """Create back-ground example for the SHAP explainer as a mode/median of features of the data to explain."""
+    median = df.median()
+    example = df.mode(dropna=False).head(1)
+    num_columns = [key for key in list(df.columns) if input_types.get(key) == "numeric"]
+
+    # Replace a numerical features' mode on their median.
+    for column in num_columns:
+        example[column] = median[column]
+
+    example = example.astype(df.dtypes)
+    return example
 
-    input_df = prepare_df(pd.DataFrame([input_data]))
+
+def explain_full(model: BaseModel, dataset: Dataset) -> np.ndarray:
+    """Perform SHAP values calculation for samples of a given dataset."""
+    # Prepare background sample to be used in the KernelSHAP.
+    background_df = model.prepare_dataframe(dataset.df, dataset.column_dtypes, dataset.target)
+    background_sample = background_example(background_df, dataset.column_types)
+
+    # Prepare input data for explanation.
+    input_df = prepare_df(dataset.df, model=model, dataset=dataset)
+
+    # Obtain SHAP explanations.
+    explainer = KernelExplainer(model.predict_df, background_sample, feature_names=input_df.columns, keep_index=True)
+    shap_values = explainer.shap_values(input_df, silent=True)
+    return shap_values
+
+
+def explain_with_shap(model: BaseModel, dataset: Dataset) -> ShapResult:
+    """Get SHAP explanation result."""
+    # Obtain SHAP explanations.
+    shap_values = explain_full(model, dataset)
+
+    # For classification, take SHAP of prediction with the highest probability.
+    if model.is_classification:
+        shap_values = _get_cls_prediction_explanation(model, dataset, shap_values)
```

Comment:
let's keep this a general function:
- I'd introduce a boolean `only_highest_proba` to `explain_with_shap` and `ShapResult` (to allow the user to choose) as follows:
- I'd introduce `model_type` in `Shapresult`
```py
def explain_with_shap(model: BaseModel, dataset: Dataset, only_highest_proba=True):
  # ...
  if only_highest_proba and model.is_classification:
    shap_values = _get_highest_proba_shap_values(model, dataset, shap_values)

  return ShapResult(shap_explanations, feature_types, feature_names, model.meta.model_type, only_highest_proba)
```
so that:
```python
@dataclass
class ShapResult:
    explanations: Explanation = None
    feature_types: dict = None
    feature_names: list = None
    model_type: ModelType = None
    only_highest_proba: bool = True
```
and then I would check `if self.only_highest_proba and self.model_type == SupportedModelTypes.CLASSIFICATION ` inside `to_wandb` to filter out the proba, and `elif not self.only_highest_proba and self.model_type == SupportedModelTypes.CLASSIFICATION ` I would raise an error saying that we only support SHAP logging with `only_highest_proba=True` for `classification` models.

Reply:
I do not understand, why we need to filter probe once again inside ShapResult().to_wandb(), if we already did this before, in the 'explain_with_shap'

Reply:
yeah I wrote this at the beginning and forgot to update it, of course, the filtering is needed once in `explain_with_shap`

Reply:
thanks!

Reply:
Done